hw/acpi/piix4.c

```diff
@@ -550,7 +550,13 @@ static void piix4_pm_realize(PCIDevice *dev, Error **errp)
     if (pcms->fw_cfg) {
         qemu_register_reset(piix4_reset, s);
     } else {
-        piix4_reset(s);
+        /*set the pm base in qemu side*/
+        pci_set_long(d->config + PIIX_PMBASE, ACPI_PM_BASE | 0x01);
+        /* enable PM io space */
+        pci_set_byte(d->config + PIIX_PMREGMISC, 0x01);
+
+        pm_io_space_update(s);
+        qemu_register_reset(piix4_reset_nofw, s);
```

Comment:
PM stuff should be done before registering a reset function? or can we move the PM stuff inside the reset function?

Reply:
@devimc Yes, in nofw case, guest OS initializes ACPI which needs base address but qemu doesn't give it. (In firmware case, firmware will do that. ) Thus, guest dmesg will show 
```
"[    0.000000] ACPI BIOS Error (bug): Required FADT field Pm1aEventBlock has zero address and/or length: 0x0000000000000000/0x4 (20180531/tbfadt-605)" 
```
Actually the patch <6ba2bfbe> just fixes this issue, but we still need register a callback to set the "acpi_get_bsel" property so hotplug can work. So in my commit, I do the pm stuff for nofw and also register the callback with a new function named "piix4_reset_nofw". It doesn't change anything in firmware case.

Reply:
@liujing2 ok, sounds good

can we move the PM stuff and have a `piix4_reset_nofw` function similar to this?

```C
static void piix4_reset_nofw(void *opaque)
{
    PIIX4PMState *s = opaque;
    PCIDevice *d = PCI_DEVICE(s);

    /*set the pm base in qemu side*/
    pci_set_long(d->config + PIIX_PMBASE, ACPI_PM_BASE | 0x01);
    /* enable PM io space */
    pci_set_byte(d->config + PIIX_PMREGMISC, 0x01);
    pm_io_space_update(s);

    acpi_pcihp_reset(&s->acpi_pci_hotplug);
}
```

Reply:
@devimc  I know what you mean. Let me introduce the running order. Firstly, the PM stuff needs to be done immediately at piix4_pm_realize() for guest initializing ACPI tables, so this part could not be part of the callback(when callback is called, it is too late). Secondly, if acpi_pcihp_reset() is done at the same time, acpi_set_bsel will failed to set the property because qbus_is_hotpluggable(BUS(bus)) return false. The hotplug_handler is set in piix4_pm_machine_ready that is after piix4_pm_realize. So it has to be put into a callback and waiting to be called in qemu_devices_reset().  This is a little complex and hope my words clear :)

For your consideration, I think maybe I can put the PM stuff in another function to make it beautiful? Which do you like?